[AArch64][SME] Fix accessing the emergency spill slot with hazard padding

MacDue opened the pull request:
This patch fixes an issue where when hazard padding was enabled locals, including the emergency spill slot, could not be directly addressed.

Generally, this is fine, we can materialize the constant offset in a scratch register, but if there's no register free we need to spill, and if we can't even reach the emergency spill slot then we fail to compile.

This patch fixes this by ensuring that if a function has variable-sized objects and is likely to have hazard padding we enable the base pointer. Then if we know a function has hazard padding, place the emergency spill slot next to the BP/SP, to ensure it can be directly accessed without stepping over any hazard padding.

MacDue committed:
[AArch64][SME] Fix accessing the emergency spill slot with hazard padding

This patch fixes an issue where when hazard padding was enabled locals,
including the emergency spill slot, could not be directly addressed.

Generally, this is fine, we can materialize the constant offset in a
scratch register, but if there's no register free we need to spill, and
if we can't even reach the emergency spill slot then we fail to compile.

This patch fixes this by ensuring that if a function has variable-sized
objects and is likely to have hazard padding we enable the base pointer.
Then if we know a function has hazard padding, place the emergency spill
slot next to the BP/SP, to ensure it can be directly accessed without
stepping over any hazard padding.

MacDue commented:
Reduced godbolt case for the crash https://godbolt.org/z/eafscqW1c (using handwritten MIR as it's simpler, but the actual crash was found when marking a real benchmark function as streaming-compatible).

efriedma-quic reviewed (commented):
Instead of forcing a base pointer, we could relocate the emergency spill slot so it's part of the callee-save region.  But I guess if you do that's, it's expensive to refer to local variables, even if it's still technically possible.  Not sure.  Having two variable-sized regions on the stack makes the stack layout complicated.

MacDue commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -632,14 +633,26 @@ bool AArch64RegisterInfo::hasBasePointer(const MachineFunction &MF) const {
+    // Since hasBasePointer() is called before we know if we have hazard padding
+    // or an emergency spill slot we need to enable the basepointer
+    // conservatively.
+    if (AFI->hasStackHazardSlotIndex() ||
```
It's a much cheaper check than parsing all the SME attributes and `hasBasePointer()` is called again repeatedly after we know we have hazard padding (e.g. when resolving frame indexes).

efriedma-quic replied:
If parsing the SME attributes is expensive, maybe we should store the parsed SME attributes in AFI?

MacDue replied:
That'd probably be sensible :+1: (the parsed attributes are just a single `int`), I can do that as a follow-up NFC.

MacDue committed:
Don't enable BP when there's definitely no hazard padding

Maybe we should (as some of these cases look better), but this was
not intentional I missed a check.

efriedma-quic reviewed (approved):
LGTM

MacDue commented:
> Instead of forcing a base pointer, we could relocate the emergency spill slot so it's part of the callee-save region. But I guess if you do that's, it's expensive to refer to local variables, even if it's still technically possible.

Hmm, that'd probably require a little more rework, and given the base pointer is likely to be beneficial in this case, I don't feel the need to avoid it.

MacDue commented:
/cherry-pick b5cf030
